Nodejs/Product/Nodejs/SharedProject/Wpf/NotificationTextBox.cs

```diff
@@ -0,0 +1,103 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Windows.Automation.Peers;
+using System.Windows.Automation.Provider;
+using System.Windows.Controls;
+
+namespace Microsoft.NodejsTools.SharedProject.Wpf
+{
+    internal class NotificationTextBox : TextBox
+    {
+        // This control's AutomationPeer is the object that actually raises the UIA Notification event.
+        private NotificationTextBoxAutomationPeer notificationTextBoxAutomationPeer;
+
+        // Assume the UIA Notification event is available until we learn otherwise.
+        // If we learn that the UIA Notification event is not available, no instance
+        // of the NotificationTextBox should attempt to raise it.
+        private static bool notificationEventAvailable = true;
+        public bool NotificationEventAvailable
+        {
+            get
+            {
+                return notificationEventAvailable;
+            }
+            set
+            {
+                notificationEventAvailable = value;
+            }
+        }
+
+        protected override AutomationPeer OnCreateAutomationPeer()
+        {
+            this.notificationTextBoxAutomationPeer = new NotificationTextBoxAutomationPeer(this);
+
+            return this.notificationTextBoxAutomationPeer;
+        }
+
+        public void RaiseNotificationEvent(string notificationText, string notificationGuid)
+        {
+            // Only attempt to raise the event if we already have an AutomationPeer.
+            if (this.notificationTextBoxAutomationPeer != null)
```

Comment:
Is any synchronization required here?

Reply:
I don't think so but I might be missing something. Is there a specific scenario you think might affect this?

Reply:
Is there some reason to believe the calls to `OnCreateAutomationPeer` and `RaiseNotificationEvent` won't overlap?

Reply:
Offline you pointed out that it doesn't actually matter if they overlap.